header.c

```diff
@@ -528,7 +528,7 @@ static int sam_hrecs_vadd(sam_hrecs_t *hrecs, const char *type, va_list ap, ...)
     sam_hrec_type_t *h_type;
     sam_hrec_tag_t *h_tag, *last=NULL;
     int new;
-    khint32_t type_i = (type[0]<<8) | type[1], k;
+    khint32_t type_i = ((unsigned)type[0]<<8) | type[1], k;
```

Comment:
This code appears eight times in _header.c_. Better to abstract it into `static inline unsigned int hdrtype2int(const char *type)` and fix seven sets of latent sign extension bugs at once!